### PR TITLE
[Bug fix] Build the reference rope from the config, not the model

### LINKED ISSUE #54516: [model]: prefill reference rope is built in bf16, so PCC decays with position
### Component / Area

model — `models/demos/deepseek_v3_d_p` (prefill reference path)

### Issue Type

Bad Outputs

**Found during:** Mistral Small 4 119B prefill bring-up (#53688). The bug is in shared DeepSeek-family prefill code and affects every MLA variant's long-context validation, not just Mistral.

## Observed

The torch reference used for prefill PCC validation gets its rotary table from an already-instantiated model. The reference is instantiated in bfloat16, so `inv_freq` is a bf16 **buffer** — 0.75 where the true value is 0.7498942 — and the later `.float()` upcasts a number that has already lost its low bits. Evaluating "in fp32" changes nothing.

The residue is a ~4.4e-4 per-dimension frequency error, i.e. a phase error growing **linearly with position**. It hits only the rotary half of the KVPE cache, and attention output stays ~0.99 because q and k share the same wrong table and the error cancels in q·k — so it reads exactly like a device rope kernel bug and is not one.

Measured at mesh-8x4, KVPE PE part: **0.997590 at 1024 tokens, 0.946176 at 5120**.

## Expected

The reference rotary table is fp32 regardless of the dtype the reference model was built in, so reference-vs-device PCC is flat in sequence length rather than decaying with position.

## Steps to Reproduce

1. `pytest models/demos/deepseek_v3_d_p/tests/test_prefill_block.py` for any MLA variant, at 1024 and then 5120 tokens.
2. Compare the KVPE PE-part PCC at the two lengths: it degrades with position (0.9976 → 0.9462).

**Input data:** any MLA variant (DeepSeek-V3, Kimi-K2, GLM-5.2, Mistral Small 4) and its HF checkpoint.
**Frequency:** Always; the error scales with position, so short runs look clean.

## Software Versions

- tt-metal main `58342a6f63d` (2026-08-26) — bug present at this commit
- `transformers == 5.12.1`, `torch == 2.11.0+cpu` (repo pins)
- Ubuntu 22.04.5 LTS

## Hardware Details

- Product: Blackhole
- Card/System: Galaxy, 32-chip (`bh-glx-110-a04u02`)

## Regression

Not a regression — the reference has been built from the model object since it was introduced. It only became visible at long context.

## Priority

P2

## Impact

Long-context prefill accuracy validation for MLA variants whose reference layer takes `position_embeddings` (transformers >= 5; Mistral Small 4 is the measured case). Worse than a plain failure: it points the investigation at the device.

**Wider than the measured case.** `create_hf_model` casts the whole reference to bf16, and that cast converts the non-persistent `inv_freq` buffer — verified: 0.7498942 becomes 0.75, and `.float()` returns 0.75, not the original. So variants whose vendored layer computes rope *internally* (DeepSeekV3, Kimi, GLM) read the same corrupted table. That path is not addressed by the fix for this issue, which supplies an fp32 table only where the layer signature accepts one. Its practical PCC impact has not been measured; worth a follow-up rather than an assumption.

### PR DESCRIPTION
**Pull request by kmabeeTT**
### PR Category
Bug fix

### Summary

Fixes #54516.

Rebuild the reference rotary table from the **config** in fp32 rather than deep-copying it off an instantiated model, so a bf16 reference no longer yields a bf16 `inv_freq`. Built once per run rather than per layer — it depends only on the positions, and each build would otherwise construct a rotary module and a full fp32 copy of the hidden states (~251 MB at seq 15360).

The table is built only when a layer's **signature** actually takes `position_embeddings`. A vendored reference such as DeepSeekV3 computes rope internally and exposes no top-level `rotary_emb`, so building it unconditionally asserted on precisely the models that never needed one.

Also adds mscale to the chunked reference cache key — it scales softmax by `mscale**2`, so toggling it served a stale reference — and a generic `reference_rotary_cls` adapter property.

Split out of #53747 per review.

### Notes for reviewers

- Block test at mesh-8x4, KVPE PE part: **0.997590 → 0.999900** at 1024 tokens, **0.946176 → 0.999899** at 5120. The result is now *flat in sequence length*, which is the signature that the position-dependent term is gone rather than merely under a threshold.
- Ruled out along the way, in case the same symptom recurs: a YaRN mismatch (the DeepSeek and Mistral tables are bit-identical for this config), the half-split/interleaved permutation (PCC 1.0 on CPU), and bf16 cos/sin (four orders of magnitude too small).
- `tests/torch/test_reference_rope.py` is host-only (no device, no weights). It asserts the rebuilt fp32 table matches an independent one at position 40000 **and** that the old `deepcopy(bf16).float()` path is wrong there by >1e-3 — without that second assertion the test would pass whether or not the fix is present.
- Rebased onto `58342a6f63d`.

### Testing

Local, on a 32-chip Blackhole galaxy, on a branch carrying all five split fixes on this PR's base:

- `tests/torch/test_reference_rope.py` — host-only, 4 passed; asserts the rebuilt fp32 table matches an independent one at position 40000 **and** that the pre-fix `deepcopy(bf16).float()` path is wrong there by >1e-3
- Mistral Small 4 prefill block, 8x4 — 6 passed; KVPE PE part **0.999952** (pre-fix: 0.946 at this length), and flat in sequence length rather than decaying
- Blaze dispatched on this branch, `test-type=prefill_block,glm_prefill_block,kimi_prefill_block,glm_chunked_accuracy,minimax_prefill_kv_pcc`: [run 33043223403](https://github.com/tenstorrent/tt-metal/actions/runs/33043223403) (passing)
- These jobs were green on `3900b5ea463`, one commit below this base. The overall run there is red only from `high_bw_all_gather CCL perf + accuracy`, a pre-existing flake outside this selection.
- All five together on `kmabee/issue_54515-54519_integration` (base `58342a6f63d`): Blaze [`test-type=all`](https://github.com/tenstorrent/tt-metal/actions/runs/33094428995) green 44/44, [(Blackhole) e2e](https://github.com/tenstorrent/tt-metal/actions/runs/33094442486) 39/40.
- The one red is not from these changes: the DeepSeek perf threshold has 0.15% headroom against 0.41% machine variance ([clean-base control on the same machine passed](https://github.com/tenstorrent/tt-metal/actions/runs/33115646268)).

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:kmabee/issue_54516_fix-mla-reference-rope)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:kmabee/issue_54516_fix-mla-reference-rope)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:kmabee/issue_54516_fix-mla-reference-rope)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:kmabee/issue_54516_fix-mla-reference-rope)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:kmabee/issue_54516_fix-mla-reference-rope)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:kmabee/issue_54516_fix-mla-reference-rope)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:kmabee/issue_54516_fix-mla-reference-rope)

### CODE REVIEW

File: models/demos/deepseek_v3_d_p/utils/transformer_helpers.py [hunk 1]
```diff
@@ -579,14 +640,18 @@ def load_and_compute_layer_by_layer(
...
-                    attention_mask=attention_mask,
-                    position_ids=position_ids,
-                    past_key_value=ref_cache,
-                    use_cache=True,
-                )
-                h_ref = layer_out[0]
+                layer_kwargs = {
+                    "attention_mask": attention_mask,
+                    "position_ids": position_ids,
+                    "past_key_value": ref_cache,
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply models/demos/deepseek_v3_d_p/utils/transformer_helpers.py]: Fixed in 81a50010eba — the call site now binds the cache kwarg from the layer's own signature, same as `decoder_layer_kwargs`.

File: models/demos/deepseek_v3_d_p/tests/reference_runners.py [hunk 2]
```diff
@@ -119,14 +120,32 @@ def run_reference_mla(
...
+        assert rotary_cls is not None, (
+            f"{type(attn).__name__} requires position_embeddings but {variant.name!r} exposes no "
+            "reference_rotary_cls to build (cos, sin) from"
         )
+        # Built from the CONFIG, not from the model: a reference instantiated in bf16 carries a bf16
+        # `inv_freq` buffer, and .float() cannot recover the lost bits. The resulting ~4.4e-4
+        # frequency error is a phase error that grows with position.
+        rotary = rotary_cls(config=config).float()
+        with torch.no_grad():
+            cos, sin = rotary(hidden_states.float(), position_ids)
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply models/demos/deepseek_v3_d_p/tests/reference_runners.py]: Fixed in 81a50010eba — passes a `[:1, :1]` view, matching `reference_rope`.

File: models/demos/deepseek_v3_d_p/utils/chunked_prefill_utils.py [hunk 3]
```diff
@@ -47,14 +47,24 @@ def _ref_cache_key(config, weights, hidden_2d) -> str:
...
     cache dir is safe.
     """
     h = hashlib.sha1()
-    for field in ("num_attention_heads", "rms_norm_eps", "mla_use_nope", "mla_use_output_gate", "rope_scaling"):
+    for field in (
+        "num_attention_heads",
+        "rms_norm_eps",
+        "mla_use_nope",
+        "mla_use_output_gate",
+        "mla_disable_yarn_mscale",
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply models/demos/deepseek_v3_d_p/utils/chunked_prefill_utils.py:65]: Added in 81a50010eba: `test_mscale_toggle_invalidates_the_chunked_reference_cache` holds weights and hidden state constant, asserts the key changes when only that flag flips, and that it is stable otherwise.

File: models/demos/deepseek_v3_d_p/utils/transformer_helpers.py [hunk 4]
```diff
@@ -213,6 +214,56 @@ def create_hf_model(variant, config, num_layers, n_routed_experts=None):
...
+    """
+    rotary = getattr(hf_model, "rotary_emb", None)
+    assert rotary is not None, f"{type(hf_model).__name__} exposes no rotary_emb to build rope from"
+    # Check the signature rather than catching: a bare `except` would also swallow a genuine
+    # construction failure and silently fall back to the bf16 buffer this function exists to avoid.
+    rotary_cls = type(rotary)
+    if "config" in inspect.signature(rotary_cls.__init__).parameters:
+        rotary_f32 = rotary_cls(config=hf_model.config).float()
+    else:  # a rotary that predates `config=`; the bf16 buffer is then unavoidable
+        rotary_f32 = deepcopy(rotary).float()
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply models/demos/deepseek_v3_d_p/utils/transformer_helpers.py:243]: Addressed in 678784d5f2e, as a warning rather than a raise. The four config-less rotaries under `reference/` are exactly the DeepseekV3/Kimi ones; their vendored layers compute rope internally so they never reach this branch today, but raising would turn that reachability assumption into a latent break for the two variants with the most CI coverage. The warning names the class and removes the silence, which was the actual problem.

File: models/demos/deepseek_v3_d_p/tests/torch/test_reference_rope.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+import torch
+
+from models.demos.deepseek_v3_d_p.utils.transformer_helpers import (
+    layer_wants_position_embeddings,
+    reference_position_embeddings,
+    reference_rope,
+)
+
+
+def test_reference_rope_is_fp32_even_when_the_model_is_bf16():
```
[the remark this answers is not in the document]
Comment (kmabeeTT) [reply models/demos/deepseek_v3_d_p/tests/torch/test_reference_rope.py:25]: Correct, and it is a known gap rather than an oversight: nothing under `tests/torch/` is gated today. The mechanism exists (`cpu_medium` + `fabric_cpu_only_unit_tests.yaml`) but `time_budget.yaml` grants `cpu_medium` only to `scaleout`, and these tests are `team: models`. Wiring that up needs a shared-config change, so it is deliberately out of scope here. Leaving open for the reviewer.

**Comment by kmabeeTT** [comment]
Testing section updated with CI links. Targeted Blaze run on this branch is green, and all five split fixes were also raised together on a combined branch (`kmabee/issue_54515-54519_integration`, same base) to check they compose: Blaze `test-type=all` green 44/44, plus (Blackhole) e2e. The one e2e red there is a perf threshold with 0.15% headroom against 0.41% machine variance — a clean-base control on the same machine passed, so it is not from these changes.

**Review by jjovicicTT — APPROVED** [review]
